### ISSUE TITLE
Avoid And and Or argument casting

### ISSUE THREAD

**Issue by Vatavuk** [yegor256/cactoos#792]
In order to use `And`/`Or` without compilaton issues we have to cast first argument like in this example:
```java
final List<Integer> list = new LinkedList<>();
new Or(
    (Proc<Integer>) list::add,
    2, 3, 4
).value();
```
I propose to parametrize those classes to simplify usage:
```java
final List<Integer> list = new LinkedList<>();
new Or<>(
    list::add,
    2, 3, 4
).value();
```

**Comment by 0crat** [comment]
@llorllale[/z](https://www.0crat.com/u/llorllale) please, pay attention to this issue

**Comment by llorllale** [comment]
@Vatavuk those ctors that accept `Proc` are there for convenience really. Semantically speaking, it makes no sense to pass `void` types to these boolean algebraic types.

I think we need something else here that will *consume* the iterator/iterable with the provided action. Something similar to #476.

**Comment by Vatavuk** [comment]
@llorllale The problem is the same for ctor that takes `Func`.
Example from one project of mine:
```java
retur new Or(
    (Func<Wire, Boolean>) input -> input.isActive(context),
    this.wires
).value();
```
This casting introduces noise

**Comment by llorllale** [comment]
@Vatavuk yes, I agree. The problem is between those two overloaded versions of the ctors (Proc/Func). What I'm saying is, perhaps we should get rid of the overload that accepts the `Proc` and offer some other alternative.

**Comment by Vatavuk** [comment]
@llorllale I didn't know that we can specify type in lambda input and avoid casting.
The above example can look like this:
```java
retur new Or(
    (Wire input) -> input.isActive(context),
    this.wires
).value();
```
I think that this is ok. 
I'm not sure about removing those ctors because this functionallity fits naturaly to `Or` object.
Do you agree to close this issue?

**Comment by llorllale** [comment]
@Vatavuk if you think it's ok then you may close your own issue :)

**Comment by 0crat** [comment]
The job is not in WBS, won't close the order
